### PR TITLE
Revisit design for third-party drivers

### PR DESCRIPTION
**Pull request by ogayot**
Marking this a draft since more work would need to be done with the answers support. Currently, the source screen is skipped from the integration tests and I haven't figured out why/how.

A new checkbox is shown in the source screen, at the beginning of the installation. This new checkbox specifies if the user "is interested in third party drivers" or wants to skip out the third-party drivers screen completely.

source

The drivers views have been updated to show slightly different wording when running an offline install vs an online install.

no-drivers-offline
install

The nature of the change introduces some coupling between the "source" components and the "drivers" components. Currently the state of the  "Search for drivers" is contained in the "source" model ; which forces the "drivers" client controller to make HTTP calls to the "source" controller. There might be better solutions?

### CODE REVIEW

**Comment by ogayot** [comment]
Answers support is now implemented.
I had to adjust the way the "Source" controller was handled in automated tests.
Before, we would not consider the controller interactive if only one source was found. Now, we do since we have other settings to show. Also, we pass the sources catalog to the integration tests.

File: subiquity/server/controllers/drivers.py [hunk 1]
```diff
@@ -74,6 +74,12 @@ def start(self):
     async def _list_drivers(self, context):
         with context.child("wait_apt"):
             await self._wait_apt.wait()
+        # At this point, the user can't go back all the way to the source
+        # screen to enable/disable the "search drivers" checkbox.
```
Comment (mwhudson) [inline subiquity/server/controllers/drivers.py:78]: Er. Why is that? Maybe extend the comment a bit. One day (maybe soon!) we'll allow the user to retry apt configuration if it fails, will that invalidate this comment?

**Review by mwhudson — CHANGES_REQUESTED** [review]
One question, one requested change, otherwise looking good!

(Sorry for taking so long to get to this)

File: subiquity/server/controllers/drivers.py (hunk 1 above)
Comment (ogayot) [reply subiquity/server/controllers/drivers.py:78]: Well, `self._wait_apt.wait()` unblocks when we receive the `APT_CONFIGURED notification. 
Currently, we configure APT (and then broadcast the `APT_CONFIGURED` notification) after the user has "confirmed destructive changes" (i.e., finished partitioning). From the GUI standpoint, there is no going back beyond that point _yet?_. Also, by the time APT_CONFIGURED is received, the "source" is already mounted so we don't want the user to change it. I'll add some details to the comment.

Regarding the ability to retry the APT configuration, I'd still assume we wouldn't allow the user to change things in the sources screen, so it wouldn't invalidate the comment.

The `_list_drivers` task will probably need to be executed once more if we change the APT configuration though, but that's a different story.

File: subiquity/client/controllers/drivers.py [hunk 2]
```diff
@@ -31,10 +31,15 @@ class DriversController(SubiquityTuiController):
...
-        response: DriversResponse = await self.endpoint.GET()
-        if not response.drivers and response.drivers is not None:
+        source_endpoint = self.app.client.source
+        source_response = await source_endpoint.GET()
+
+        if not source_response.search_drivers:
             raise Skip
-        return DriversView(self, response.drivers, response.install)
+
+        response: DriversResponse = await self.endpoint.GET()
```
[the remark this answers is not in the document]
Comment (ogayot) [reply subiquity/client/controllers/drivers.py:40]: So.... I didn't want to introduce too much coupling between the source and the drivers controllers on the server side but I think this is justified in this case. I will update this.

**Comment by ogayot** [comment]
I've just realized that the autoinstall of third-party drivers does not work anymore because:
1. `search_drivers` is `False` by default
2. there is no autoinstall support for `source` (we need to add support for it)

Furthermore, we need to make sure `search_drivers` defaults to `True` in autoinstall context, so that we are backward compatible with existing autoinstall configurations that only have `drivers: install: true`

**Comment by ogayot** [comment]
Rebased on top of master to resolve conflict.
Also implemented autoinstall for the `source` controller.

**Review by mwhudson — APPROVED** [review]
Thanks for the changes!

**Comment by mwhudson** [comment]
> 2\. there is no autoinstall support for `source` (we need to add support for it)

Er oops. I guess we should add support for selecting the source via autoinstall too! (Not in this branch!)

File: subiquity/server/controllers/drivers.py (hunk 1 above)
Comment (mwhudson) [reply subiquity/server/controllers/drivers.py]: Thanks, I just wanted to make sure we're not backing ourselves into a corner!

File: subiquity/client/controllers/drivers.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mwhudson) [reply subiquity/client/controllers/drivers.py]: The way most or at least many of the "GET /$endpoint" methods return a type that is only used by that method which then mostly gets passed on to the view object feels a bit verbose and clunky but I'm not sure what would be an improvement. I _guess_ you could come up with some scheme that has the view describe what information you need from the server and metaprogram your way to a concrete type from that but .... nah sounds pretty silly (and also wouldn't help the desktop installer at all!)

[the remark this answers is not in the document]
Comment (ogayot) [reply subiquity/client/controllers/drivers.py]: Yeah, I don't love it either. Unfortunately, at the moment I don't have any good suggestion to make this better though :upside_down_face:

**Review by dbungert — APPROVED** [review]

**Pull request merged**